tools/distrib/python/xds_protos/setup.py

```diff
@@ -0,0 +1,145 @@
+
+
+# Compile xDS protos
+def has_grpc_service(proto_package_path: str) -> bool:
+    return proto_package_path.startswith('envoy/service')
+
+
+def compile_protos(proto_root: str, sub_dir: str = '.') -> None:
+    for root, _, files in os.walk(os.path.join(proto_root, sub_dir)):
+        proto_package_path = os.path.relpath(root, proto_root)
+        if proto_package_path in EXCLUDE_PROTO_PACKAGES_LIST:
+            print(f'Skipping package {proto_package_path}')
+            continue
+        for file_name in files:
+            if file_name.endswith('.proto'):
+                # Compile proto
+                if has_grpc_service(proto_package_path):
+                    return_code = protoc.main(COMPILE_BOTH +
+                                              [os.path.join(root, file_name)])
+                    add_test_import(proto_package_path, file_name, service=True)
+                else:
+                    return_code = protoc.main(COMPILE_PROTO_ONLY +
+                                              [os.path.join(root, file_name)])
+                    add_test_import(proto_package_path,
+                                    file_name,
+                                    service=False)
+                if return_code != 0:
+                    raise Exception('error: {} failed'.format(COMPILE_BOTH))
+
+
+compile_protos(XDS_PROTO_ROOT)
+compile_protos(UDPA_PROTO_ROOT)
+# We don't want to compile the entire GCP surface API, just the essential ones
+compile_protos(GOOGLEAPIS_ROOT, os.path.join('google', 'api'))
+compile_protos(GOOGLEAPIS_ROOT, os.path.join('google', 'rpc'))
+compile_protos(GOOGLEAPIS_ROOT, os.path.join('google', 'longrunning'))
+compile_protos(GOOGLEAPIS_ROOT, os.path.join('google', 'logging'))
+compile_protos(GOOGLEAPIS_ROOT, os.path.join('google', 'type'))
+compile_protos(VALIDATE_ROOT, 'validate')
+compile_protos(OPENCENSUS_PROTO_ROOT)
+
+
+# Generate __init__.py files for
+def create_init_file(path: str) -> None:
+    f = open(os.path.join(path, "__init__.py"), 'w')
+    f.close()
+
+
+create_init_file(WORK_DIR)
+for root, _, _ in os.walk(os.path.join(WORK_DIR, 'envoy')):
+    create_init_file(root)
+
+# Generate test file
+with open(os.path.join(WORK_DIR, TEST_FILE_NAME), 'w') as f:
+    f.writelines(TEST_IMPORTS)
+
+# Use setuptools to build Python package
+CLASSIFIERS = [
+    'Development Status :: 3 - Alpha',
+    'Programming Language :: Python',
+    'Programming Language :: Python :: 2',
+    'Programming Language :: Python :: 3',
+    'License :: OSI Approved :: Apache Software License',
+]
+setuptools.setup(
+    name='xds-protos',
+    version='0.0.1',
+    packages=setuptools.find_packages(where=".", exclude=[TEST_FILE_NAME]),
+    description='Generated Python code from envoyproxy/data-plane-api',
+    author='The gRPC Authors',
+    author_email='grpc-io@googlegroups.com',
+    url='https://grpc.io',
+    license='Apache License 2.0',
```

Comment:
No dependency on `protobuf`,`grpcio-tools`, or `grpcio`?

Reply:
Good catch! According to the `setuptools` doc, `install_requires` is:

> install_requires is a setuptools setup.py keyword that should be used to specify what a project minimally needs to run correctly.

So, I added `grpcio` and `protobuf`.

Reply:
You'll also need `grpcio-tools` in `setup_requires` or supplied as specified by [PEP 518](https://www.python.org/dev/peps/pep-0518/). You reference it in several places in the build.

Reply:
Done. `grpcio-tools` added as `setup_requires`.